Allow for missing install_dir in install_data()

whot opened the pull request:
The documentation doesn't require it and the interpreter code works around the
possibility of it being None. The ninja backend code however fails with

File "/home/whot/code/meson/mesonbuild/backend/ninjabackend.py", line 796, in generate_data_install
    dstabs = os.path.join(subdir or None, plain_f)
File "/usr/lib64/python3.6/posixpath.py", line 78, in join
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType

If install_dir is missing, default to '.'

jpakkane commented:
The tests fail due to a missing file. A forgotten `git add` perhaps?

That being said perhaps we should just make a missing `install_dir` an error instead? Is there a use case for installing to `prefix` that would justify making it the default`?

jpakkane commented:
Ping.

whot commented:
sorry about the silence, I had that fixed but the windows cases kept failing and then it fell off the radar. Rebased and fixed now, most importantly: install_data now goes into $datadir instead of just $prefix, I think that makes more sense. Problem is the windows cases are still failing, but i'm not quite sure what's happening there: `Reason: Extra file usr/share/somefile.txt found.` is the failure. It installs, but something isn't right here, this could be some underlying issue? Any hints are appreciated.

QuLogic commented:
You need to update the `installed_files.txt` also.

whot commented:
sometimes the solution is so close... :) @QuLogic: thanks!

updated and force-pushed, ready for merging now

nirbheek commented:
So essentially the existing behaviour was that `install_data()` would error out with a traceback if you didn't specify `install_dir:`? Maybe the real fix is to raise an `InvalidArguments` exception which will print a readable error message. Unless you can think of a use-case which makes installing into `/usr/share` a good default?

whot commented:
well, come to think of it - no. But installing into `/usr/share/packagename` would be a good default, I think.

nirbheek commented:
That sounds good, please also update the reference manual with a note about the behaviour change, and also mention that it was `*(Added 0.43.0)*`.

whot commented:
updated with the doc changes and release notes included

jpakkane commented:
I rebased this and opened a new PR, so closing this one. It'll get merged now, sorry for the delay.
